### PR TITLE
cli: add method autocompletion

### PR DESCRIPTION
**Pull request by lundibundi**
Allow to specify remote interfaces that will be inspected upon connection and uses resulting api to complete calls. The syntax is as follows

`connect [<protocol>://]<host>:<port> <application name>  [ <interface> [ ... ] ]`

### CODE REVIEW

File: tools/cli.js [hunk 1]
```diff
@@ -24,9 +24,17 @@ const log = (msg) => {
...
-  if (!input) return completions;
-  return completions.filter(c => c.startsWith(input));
+function complete(input, completions, forbidden = []) {
+  if (!forbidden && !input) return completions;
+  return completions.filter(
+    c => !startsWithAny(c, forbidden) && (!input || c.startsWith(input)));
+}
+
+function startsWithAny(str, array) {
+  for (const el of array) {
```
Comment (aqrln) [inline tools/cli.js]: I'd just do `array.some(el => str.startsWith(el))`, but this imperative approach is fine too ;)

Comment (lundibundi) [inline tools/cli.js]: Oh, this is good, will change it in a minute.

File: tools/cli.js [hunk 2]
```diff
@@ -24,9 +24,17 @@ const log = (msg) => {
   if (userInput) rl.write(userInput);
 };
 
-function complete(input, completions) {
-  if (!input) return completions;
-  return completions.filter(c => c.startsWith(input));
+function complete(input, completions, forbidden = []) {
+  if (!forbidden && !input) return completions;
+  return completions.filter(
```
Comment (belochub) [inline tools/cli.js]: As far as I understand, the result of this `filter` will always have finite number of possible values on every successful connection with interface inspection, and it does not depend on any user input after connection, which means it can be optimized by computing possible values on connection establishment (after inspection) and using them afterwards.

File: tools/cli.js [hunk 3]
```diff
@@ -141,7 +160,19 @@ commandProcessor.event.help = () => (
   'event <interfaceName> <eventName> [ <arg> [ , ... ] ]'
 );
 
-commandProcessor.connect = (protocol, host, port, appName, callback) => {
+function filterApiCompletions(rawApi) {
+  const api = {};
+  const forbidden = ['_', 'domain'];
+  Object.keys(rawApi).map((int) => {
```
Comment (aqrln) [inline tools/cli.js]: Can you please use `forEach` here so that the intention is clear?

**Comment by aqrln** [comment]
@belochub does this LGTY after the last update?

File: tools/cli.js [hunk 4]
```diff
@@ -29,7 +29,14 @@ function complete(input, completions) {
...
 
-function tryComplete(input, completer) {
+function tryComplete(input, completions) {
+  if (!completions) return input;
+  completions = complete(input, completions);
+  if (completions.length === 1) return completions[0];
+  return input;
+}
+
+function tryCompleter(input, completer) {
```
Comment (aqrln) [inline tools/cli.js]: I find it very confusing to have two functions that are completely distinct yet have names that resemble each other so close.

File: tools/cli.js [hunk 5]
```diff
@@ -116,9 +124,41 @@ commandProcessor.complete = (inputs, depth) => {
 
 commandProcessor.call = (interfaceName, methodName, args, callback) => {
   if (!state.client) return callback(new Error('Not connected'));
+
+  interfaceName = tryComplete(interfaceName, Object.keys(state.api));
```
Comment (aqrln) [inline tools/cli.js]: Do we really need to complete this?

I do see a value in allowing users to type commands incompletely, that's both convenient and consistent with tools like vi. But completing method names upon Enter without requiring the user to press Tab may lead to confusing mistakes and running random commands inadvertently. I'd prefer to force users to be explicit here.

File: tools/cli.js [hunk 6]
```diff
@@ -116,9 +124,41 @@ commandProcessor.complete = (inputs, depth) => {
...
+commandProcessor.call.complete = (inputs, depth) => {
+  if (!state.api) return [[], depth];
+
+  const iface = inputs[depth++];
+  let method = inputs[depth];
+  if (method === '' && inputs[depth + 1] !== undefined) {
+    method = inputs[++depth];
+  }
+
+    // (a and b and c) or (!a and (b or c))
```
Comment (aqrln) [inline tools/cli.js]: This should probably be removed.

File: tools/cli.js [hunk 7]
```diff
@@ -116,9 +124,41 @@ commandProcessor.complete = (inputs, depth) => {
...
+  let completions = complete(iface, Object.keys(state.api));
+  if (completions.length !== 1 && method === undefined ||
```
Comment (aqrln) [inline tools/cli.js]: Can we have a comment explaining the logic behind these conditions above the `if` statement? Aside, how about slitting it into three `if`s with a duplicate body (it's short enough to do that). I'd find it more readable.

File: tools/cli.js [hunk 8]
```diff
@@ -116,9 +124,41 @@ commandProcessor.complete = (inputs, depth) => {
...
+      completions.length !== 1 && completions[0] !== iface ||
+      method === undefined && completions[0] !== iface) {
+    return [completions, depth];
+  }
+
+  // nothing after full interface name -> show help
```
Comment (aqrln) [inline tools/cli.js]: Yeah, I was speaking about comments like this one :)

File: tools/cli.js [hunk 9]
```diff
@@ -133,7 +173,19 @@ commandProcessor.event.help = () => (
   'event <interfaceName> <eventName> [ <arg> [ , ... ] ]'
 );
 
-commandProcessor.connect = (protocol, host, port, appName, callback) => {
+function filterApiCompletions(rawApi) {
+  const api = {};
+  const forbidden = ['_', 'domain'];
+  Object.keys(rawApi).forEach((int) => {
+    api[int] = Object.keys(
+      rawApi[int]).filter(c => !forbidden.some(el => c.startsWith(el)));
```
Comment (aqrln) [inline tools/cli.js]: s/`!forbidden.some`/`forbidden.all` and invert the condition inside the lambda?

File: tools/cli.js (hunk 4 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln yeah, I totally agree on this one but I couldn't come up with a better name, what's your opinion on `tryToComplete` for the previous one and this one to remain `tryCompleter`?

File: tools/cli.js (hunk 5 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln actually that may really be the case, but I'd like to at least add cli option like
`-ac` / `--advancedCompletion` that'll enable this behavior as this may be convenient for advanced users. But I guess we will better make it in different PR and for now either add hard-coded config variable or comment it out comepletely untill we add the option? @belochub?

File: tools/cli.js (hunk 6 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln surely, thanks for catching that.

File: tools/cli.js (hunk 7 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln surely, I forgot about this one and it is quite complex.

File: tools/cli.js (hunk 9 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln fine with me.

File: tools/cli.js (hunk 5 above)
Comment (aqrln) [inline tools/cli.js]: @lundibundi I'd say I'm -1 on it one way or another, but I will not block the PR if I am in the minority.

File: tools/cli.js (hunk 9 above)
Comment (aqrln) [inline tools/cli.js]: Oh, yes, I did. Thanks for correcting me. It's been about 1.5 years since I switched to Node.js from Python and I still consistently write `all` instead of `every` and `any` instead of `some` in JavaScript 😄 This works the other way around too, though. I tend to be `push`ing elements to Python lists instead of `append`ing them.

File: tools/cli.js (hunk 4 above)
Comment (aqrln) [inline tools/cli.js]: It's more of a generic question. As far as I can see, there are functions named `complete()` and `completer()` in master already, so I'd say go ahead with `tryCompleter()` in this PR for the sake of consistency. We'll need to come up with better names for all four functions in another PR though.

File: tools/cli.js (hunk 9 above)
Comment (lundibundi) [inline tools/cli.js]: btw, ain't `all` and `any` feels better, they are a bit more clear to me?

**Comment by lundibundi** [comment]
@aqrln @belochub ping.

**Comment by aqrln** [comment]
@lundibundi the CI's failing with a linter error:

```
/home/travis/build/metarhia/JSTP/tools/cli.js
  32:10  error  'tryComplete' is defined but never used  no-unused-vars
✖ 1 problem (1 error, 0 warnings)
```

File: tools/cli.js [hunk 10]
```diff
@@ -116,9 +119,48 @@ commandProcessor.complete = (inputs, depth) => {
 
 commandProcessor.call = (interfaceName, methodName, args, callback) => {
   if (!state.client) return callback(new Error('Not connected'));
+
```
Comment (aqrln) [inline tools/cli.js]: Unrelated style change?

File: tools/cli.js [hunk 11]
```diff
@@ -116,9 +119,48 @@ commandProcessor.complete = (inputs, depth) => {
...
 
+commandProcessor.call.complete = (inputs, depth) => {
+  if (!state.api) return [[], depth];
+
+  let iface = inputs[depth++];
+  let method = inputs[depth];
+  // there may be multiple spaces between interface and method names
+  // this function completes both of them so handle empty element ('')
+  // in between (just skip it)
+  if (method === '' && inputs[depth + 1] !== undefined) {
```
Comment (aqrln) [inline tools/cli.js]: Should this be `while`?

File: tools/cli.js [hunk 12]
```diff
@@ -116,9 +119,48 @@ commandProcessor.complete = (inputs, depth) => {
...
+    method = inputs[++depth];
+  }
+
+  let completions = complete(iface, Object.keys(state.api));
+
+  const fullInterfaceMatch = completions.some(el => el === iface);
+
+  // we should show completions if
+  // there is more than one completion and no method provided
+  if (completions.length !== 1 && method === undefined ||
```
Comment (aqrln) [inline tools/cli.js]: Can `completions` be an empty array? If yes, then this condition contradicts the comment above.

File: tools/cli.js [hunk 13]
```diff
@@ -286,7 +342,15 @@ lineProcessor.connect = (tokens, callback) => {
   if (appName === undefined) {
     return callback(reportMissingArgument('Application name'));
   }
-  commandProcessor.connect(scheme, host, port, appName, (err) => {
+  let interfaces = [];
+  if (args[2]) {
+    try {
+      interfaces = jstp.parse('[' + args[2] + ']');
```
Comment (aqrln) [inline tools/cli.js]: I find it a bit odd to be forced to put quotes around interface names in the `connect` command while I don't have to do that for an application name.

File: tools/cli.js (hunk 10 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln yeah, missed it while deleting call completion.

File: tools/cli.js (hunk 11 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln no, `_split` function collapses all consequent separators.

**Comment by lundibundi** [comment]
Landed in b206c0540384b43d388538960fdff5c8e894d67c.
